src/openms/source/ANALYSIS/OPENSWATH/MRMScoring.cpp

```diff
@@ -88,24 +88,26 @@ namespace OpenSwath
       return xcorr_precursor_combined_matrix_;
     }
 
+    // void MRMScoring::setValueGeneralXCorrMatrix(XCorrMatrixType& matrix, 
+    //                                 const std::vector<std::vector<double>>& intensityi, 
+    //                                 const std::vector<std::vector<double>>& intensityj)
+    // {
+    //   matrix.setValue(i, j, Scoring::normalizedCrossCorrelation(intensityi[i], intensityj[j], boost::numeric_cast<int>(intensity[i].size()), 1));
+    // }
+
     void MRMScoring::initializeXCorrMatrix(OpenSwath::IMRMFeature* mrmfeature, const std::vector<String>& native_ids)
     {
-      std::vector<double> intensityi, intensityj;
+      std::vector<std::vector<double>> intensity;
+      fillIntensityFromFeature(mrmfeature, native_ids, intensity);
       xcorr_matrix_.resize(native_ids.size(), native_ids.size());
       xcorr_matrix_max_peak_.resize(native_ids.size(), native_ids.size());
       xcorr_matrix_max_peak_sec_.resize(native_ids.size(), native_ids.size());
       for (std::size_t i = 0; i < native_ids.size(); i++)
       {
-        FeatureType fi = mrmfeature->getFeature(native_ids[i]);
-        intensityi.clear();
-        fi->getIntensity(intensityi);
         for (std::size_t j = i; j < native_ids.size(); j++)
         {
-          FeatureType fj = mrmfeature->getFeature(native_ids[j]);
-          intensityj.clear();
-          fj->getIntensity(intensityj);
           // compute normalized cross correlation
-          xcorr_matrix_.setValue(i, j, Scoring::normalizedCrossCorrelation(intensityi, intensityj, boost::numeric_cast<int>(intensityi.size()), 1));
+          xcorr_matrix_.setValue(i, j, Scoring::normalizedCrossCorrelation(intensity[i], intensity[j], boost::numeric_cast<int>(intensity[i].size()), 1));
```

Comment:
this is duplicate work, right?
each intensity[i] is normalized within `Scoring::normalizedCrossCorrelation`, but this may also be precomputed to save some time?